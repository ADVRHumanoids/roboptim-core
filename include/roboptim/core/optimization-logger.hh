// Copyright (C) 2013 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_OPTIMIZATION_LOGGER_HH
# define ROBOPTIM_CORE_OPTIMIZATION_LOGGER_HH
# include <string>

# include <boost/bind.hpp>
# include <boost/date_time/posix_time/posix_time.hpp>
# include <boost/filesystem.hpp>
# include <boost/filesystem/fstream.hpp>
# include <boost/format.hpp>
# include <boost/variant/apply_visitor.hpp>
# include <boost/variant/static_visitor.hpp>

# include <roboptim/core/config.hh>

namespace roboptim
{
  namespace detail
  {
    template <typename P>
    struct EvaluateConstraint : public boost::static_visitor<typename P::vector_t>
    {
      EvaluateConstraint (const typename P::vector_t& x)
	: x_ (x)
      {}

      template <typename U>
      typename P::vector_t operator () (const U& constraint) const
      {
	return constraint->operator () (x_);
      }

    private:
      typename P::vector_t x_;
    };

    template <typename P>
    struct JacobianConstraint
      : public boost::static_visitor<typename P::function_t::jacobian_t>
    {
      JacobianConstraint (const typename P::vector_t& x)
	: x_ (x)
      {}

      template <typename U>
      typename P::function_t::jacobian_t operator () (const U& constraint) const
      {
	return constraint->jacobian (x_);
      }

    private:
      typename P::vector_t x_;
    };


    struct ConstraintName : public boost::static_visitor<std::string>
    {
      template <typename U>
      std::string operator () (const U& constraint) const
      {
	return constraint->getName ();
      }
    };

    template <typename P>
    struct EvaluateConstraintViolation
    {
      typedef typename P::vector_t vector_t;
      typedef typename P::value_type value_type;
      typedef typename P::intervalsVect_t intervalsVect_t;
      typedef typename P::interval_t interval_t;

      EvaluateConstraintViolation
      (const std::vector<vector_t>& constraints,
       const intervalsVect_t& bounds)
        : constraints_ (constraints),
          bounds_ (bounds)
      {}

      value_type computeViolation (const value_type& x,
                                   const interval_t& x_ul) const
      {
        if (x < x_ul.first)
          return x_ul.first - x;
        else if (x_ul.second < x)
          return x - x_ul.second;
        else return 0.;
      }

      value_type uniformNorm () const
      {
        value_type norm = 0.;
        value_type cstr_viol = 0.;

        // For each multidimensional constraint
        for (size_t i = 0; i < constraints_.size (); ++i)
          {
            // For each dimension of the constraint
            for (size_t j = 0; j < constraints_[i].size (); ++j)
              {
                cstr_viol = computeViolation (constraints_[i][j],
                                              bounds_[i][j]);

                norm = std::max (norm, fabs (cstr_viol));
              }

          }
        return norm;
      }

    private:
      const std::vector<vector_t>& constraints_;
      const intervalsVect_t& bounds_;
    };
  } // end of namespace detail.

  template <typename T>
  class OptimizationLogger
  {
  public:
    typedef T solver_t;
    typedef typename solver_t::problem_t problem_t;
    typedef typename solver_t::problem_t::value_type value_type;
    typedef typename solver_t::problem_t::vector_t vector_t;
    typedef typename solver_t::problem_t::function_t::jacobian_t jacobian_t;
    typedef typename solver_t::solverState_t solverState_t;

    explicit OptimizationLogger (solver_t& solver,
				 const boost::filesystem::path& path)
      : solver_ (solver),
	path_ (path),
	output_ (),
	callbackCallId_ (0),
	firstTime_ (boost::posix_time::microsec_clock::universal_time ())
    {
      lastTime_ = firstTime_;

      // Set the callback.
      try
	{
	  solver_.setIterationCallback
            (boost::bind (&OptimizationLogger<T>::perIterationCallback,
                          this, _1, _2));
	}
      catch (std::runtime_error& e)
	{
	  std::cerr
	    << "failed to set per-iteration callback, "
	    << "many information will be missing from logs:\n"
	    << e.what () << std::endl;
	}

      // Remove old logs.
      boost::filesystem::remove_all (path);
      boost::filesystem::create_directories (path);

      output_.open (path / "journal.log");

      // Display banner.
      output_
	<< std::string (80, '*') << iendl
	<< "RobOptim Optimization Logger" << iendl
	<< " - current time: "
	<< boost::posix_time::to_iso_extended_string (firstTime_) << "Z" << iendl
	<< " - roboptim-core version: " ROBOPTIM_CORE_VERSION "\n"
	<< std::string (80, '*') << iendl
	;
    }

    virtual ~OptimizationLogger ()
    {
      // Unregister the callback, do not fail if this is impossible.
      try
	{
	  solver_.setIterationCallback (typename solver_t::callback_t ());
	}
      catch (std::exception& e)
	{}

      // Get current time
      boost::posix_time::ptime t =
	boost::posix_time::microsec_clock::universal_time();

      output_
	<< std::string (80, '*') << iendl
	<< "RobOptim Optimization Logger" << iendl
	<< " - current time: "
	<< boost::posix_time::to_iso_extended_string(t) << "Z" << iendl
	<< " - total elapsed time: "
	<< (t - firstTime_) << iendl
	<< std::string (80, '*') << iendl
	;

      // Cost evolution over time.
      {
	boost::filesystem::ofstream streamCost (path_ / "cost-evolution.csv");
	streamCost << "Cost\n";
	for (std::size_t i = 0; i < costs_.size (); ++i)
	  streamCost << costs_[i] << "\n";
      }

      // Constraint violation evolution over time.
      {
        boost::filesystem::ofstream streamCstrViol
          (path_ / "constraint-violation-evolution.csv");
        streamCstrViol << "Constraint violation\n";
        for (std::size_t i = 0; i < constraintViolations_.size (); ++i)
          streamCstrViol << constraintViolations_[i] << "\n";
      }

      // X evolution over time.
      {
	boost::filesystem::ofstream streamX (path_ / "x-evolution.csv");
	if (!x_.empty ())
	  {
	    for (std::size_t i = 0; i < x_[0].size (); ++i)
	      {
		if (i > 0)
		  streamX << ", ";
		streamX << "X " << i;
	      }
	    streamX << "\n";
	    for (std::size_t nIter = 0; nIter < x_.size (); ++nIter)
	      {
		for (std::size_t i = 0; i < x_[nIter].size (); ++i)
		  {
		    if (i > 0)
		      streamX << ", ";
		    streamX << x_[nIter][i];
		  }
		streamX << "\n";
	      }
	  }
      }


      // Constraints evolution over time.
      if (!constraints_.empty ())
	for (std::size_t constraintId = 0; constraintId < constraints_[0].size ();
	     ++constraintId)
	  {
	    boost::filesystem::ofstream streamConstraint
	      (path_ / (boost::format ("constraint-%d-evolution.csv") % constraintId).str ());
	    for (std::size_t i = 0; i < constraints_[0][constraintId].size (); ++i)
	      {
		if (i > 0)
		  streamConstraint << ", ";
		streamConstraint << "output " << i;
	      }
	    streamConstraint << "\n";
	    for (std::size_t nIter = 0; nIter < constraints_.size (); ++nIter)
	      {
		for (std::size_t i = 0; i < constraints_[nIter][constraintId].size (); ++i)
		  {
		    if (i > 0)
		      streamConstraint << ", ";
		    streamConstraint << constraints_[nIter][constraintId][i];
		  }
		streamConstraint << "\n";
	      }
          }
    }

  protected:
    void perIterationCallback (const problem_t& pb,
                               const solverState_t& state)
    {
      try
	{
          perIterationCallbackUnsafe (pb, state);
	}
      catch (std::exception& e)
	{
	  std::cerr << e.what () << std::endl;
	}
      catch (...)
	{
	  std::cerr << "unknown exception" << std::endl;
	}

      // Turn is finished, update variables.
      lastTime_ =
	boost::posix_time::microsec_clock::universal_time ();
      ++callbackCallId_;
    }

    virtual
    void perIterationCallbackUnsafe
    (const typename solver_t::problem_t& pb,
     const typename solver_t::solverState_t& state)
    {
      // Create the iteration-specific directory.
      boost::filesystem::path iterationPath =
	path_ / (boost::format ("iteration-%d") % callbackCallId_).str ();
      boost::filesystem::remove_all (iterationPath);
      boost::filesystem::create_directories (iterationPath);

      // Compute intermediary values.
      // - Store X
      const typename solver_t::vector_t& x = state.x ();
      x_.push_back (x);
      // - Get current time
      boost::posix_time::ptime t =
	boost::posix_time::microsec_clock::universal_time ();
      // - Current cost
      value_type cost;
      if (!state.cost ())
        cost = pb.function ()(x)[0];
      else cost = *state.cost ();
      costs_.push_back (cost);
      // - Current constraint violation
      value_type cstrViol;
      if (!state.constraintViolation ())
        cstrViol = 0;
      else cstrViol = *state.constraintViolation ();

      // Update journal
      if (callbackCallId_ == 0)
	output_ << solver_ << iendl;

      output_
	<< std::string (80, '+') << iendl
	<< boost::format ("Callback call number: %d") % callbackCallId_ << iendl
	<< "Elapsed time since last call: " << (t - lastTime_) << iendl
	<< "- x:" << incindent << iendl
	<< x << decindent << iendl
	<< "- f(x):" << incindent << iendl
        << cost << decindent << iendl;

      // Log all data
      // x
      boost::filesystem::ofstream streamX (iterationPath / "x.csv");
      for (std::size_t i = 0; i < x.size (); ++i)
	{
	  streamX << x[i];
	  if (i < x.size () - 1)
	    streamX << ", ";
	}
      streamX << "\n";

      // cost
      boost::filesystem::ofstream streamCost (iterationPath / "cost");
      streamCost << cost << "\n";

      // constraints
      std::vector<vector_t> constraintsOneIteration (pb.constraints ().size ());
      for (std::size_t constraintId = 0; constraintId < pb.constraints ().size ();
	   ++constraintId)
	{
	  // Create local path.
	  boost::filesystem::path constraintPath =
	    iterationPath / (boost::format ("constraint-%d") % constraintId).str ();
	  boost::filesystem::remove_all (constraintPath);
	  boost::filesystem::create_directories (constraintPath);

	  // Log name
	  boost::filesystem::ofstream nameStream (constraintPath / "name");
	  nameStream << boost::apply_visitor
	    (::roboptim::detail::ConstraintName (), pb.constraints ()[constraintId])
		     << "\n";
	  // Log value
	  boost::filesystem::ofstream constraintValueStream (constraintPath / "value.csv");

	  vector_t constraintValue = boost::apply_visitor
	    (::roboptim::detail::EvaluateConstraint<problem_t> (x), pb.constraints ()[constraintId]);
	  for (std::size_t i = 0; i < constraintValue.size (); ++i)
	    {
	      constraintValueStream << constraintValue[i];
	      if (i < constraintValue.size () - 1)
		constraintValueStream << ", ";
	    }
	  constraintValueStream << "\n";
	  constraintsOneIteration[constraintId] = constraintValue;

	  // Jacobian
	  boost::filesystem::ofstream jacobianStream (constraintPath / "jacobian.csv");
	  jacobian_t jacobian = boost::apply_visitor
	    (::roboptim::detail::JacobianConstraint<problem_t> (x), pb.constraints ()[constraintId]);
	  for (std::size_t i = 0; i < jacobian.rows (); ++i)
	    {
	      for (std::size_t j = 0; j < jacobian.cols (); ++j)
		{
		  jacobianStream << jacobian.coeffRef (i, j);
		  if (j < jacobian.cols () - 1)
		    jacobianStream << ", ";
		}
	      jacobianStream << "\n";
	    }
	}
      constraints_.push_back (constraintsOneIteration);

      // constraint violation
      if (!state.constraintViolation ())
        {
          // FIXME: handle argument bounds
          ::roboptim::detail::EvaluateConstraintViolation<problem_t>
            evalCstrViol (constraintsOneIteration, pb.boundsVector ());
          cstrViol = evalCstrViol.uniformNorm ();
        }
      constraintViolations_.push_back (cstrViol);

      boost::filesystem::ofstream streamCstrViol (iterationPath / "constraint-violation");
      streamCstrViol << cstrViol << "\n";

      if (pb.constraints (). size () > 0)
        output_ << "- viol_g(x):" << incindent << iendl
                << cstrViol << decindent << iendl;

      output_ << std::string (80, '-') << iendl;
    }

  protected:
    const solver_t& solver () const throw ()
    {
      solver_;
    }
    solver_t& solver () throw ()
    {
      solver_;
    }

    boost::filesystem::path& path () const throw ()
    {
      return path_;
    }
    boost::filesystem::path& path () throw ()
    {
      return path_;
    }
    unsigned callbackCallId () const throw ()
    {
      return callbackCallId_;
    }

  private:
    solver_t& solver_;
    boost::filesystem::path path_;
    boost::filesystem::ofstream output_;
    unsigned callbackCallId_;
    boost::posix_time::ptime lastTime_;
    boost::posix_time::ptime firstTime_;

    std::vector<vector_t> x_;
    std::vector<value_type> costs_;
    std::vector<value_type> constraintViolations_;
    std::vector<std::vector<vector_t> > constraints_;
  };
} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_OPTIMIZATION_LOGGER_HH
