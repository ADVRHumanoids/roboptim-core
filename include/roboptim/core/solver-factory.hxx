// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Implementation of the Solver class.
 */

#ifndef ROBOPTIM_CORE_SOLVER_FACTORY_HXX
# define ROBOPTIM_CORE_SOLVER_FACTORY_HXX
# include <sstream>

namespace roboptim
{
  template <typename T>
  SolverFactory<T>::SolverFactory (std::string plugin, const problem_t& pb)
  throw (std::runtime_error)
    : handle_ (),
      solver_ ()
  {
    typedef solver_t* create_t (const problem_t&);

    if (lt_dlinit () > 0)
      throw std::runtime_error ("failed to initialize libltdl.");

    std::stringstream ss;
    ss << "roboptim-core-" << plugin << "-plugin";
    handle_ = lt_dlopenext (ss.str ().c_str ());
    if (!handle_)
      {
	std::stringstream sserror;
	sserror << "libltdl failed to load plug-in ``"
		<< ss.str () << "'': " << lt_dlerror ();
	throw std::runtime_error (sserror.str ().c_str ());
      }

    create_t* c =
      reinterpret_cast<create_t*> (lt_dlsym (handle_, "create"));
    if (!c)
      {
	std::stringstream sserror;
	sserror << "libltdl failed to find symbol ``create'': "
		<< lt_dlerror ();
	throw std::runtime_error (sserror.str ().c_str ());
      }
    solver_ = c (pb);

    if (!solver_)
      {
	std::stringstream sserror;
	sserror << "libltdl failed to call ``create'': "
		<< lt_dlerror ();
	throw std::runtime_error (sserror.str ().c_str ());
      }
  }

  template <typename T>
  SolverFactory<T>::~SolverFactory () throw ()
  {
    typedef void destroy_t (solver_t*);
    destroy_t* destructor =
      reinterpret_cast<destroy_t*> (lt_dlsym (handle_, "destroy"));
    if (destructor)
      {
        destructor (solver_);
        solver_ = 0;
      }

    lt_dlclose (handle_);
    lt_dlexit ();
  }

  template <typename T>
  typename SolverFactory<T>::solver_t&
  SolverFactory<T>::operator () () throw ()
  {
    assert (solver_ != 0);
    return *solver_;
  }

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_SOLVER_FACTORY_HH
