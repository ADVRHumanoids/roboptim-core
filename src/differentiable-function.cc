// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "debug.hh"

#include "roboptim/core/differentiable-function.hh"
#include "roboptim/core/indent.hh"
#include "roboptim/core/util.hh"

namespace roboptim
{
  DifferentiableFunction::DifferentiableFunction (size_type inputSize,
						  size_type outputSize,
						  std::string name)
    throw ()
    : Function (inputSize, outputSize, name)
  {
  }

  void
  DifferentiableFunction::impl_jacobian (jacobian_t& jacobian,
					 const argument_t& argument)
    const throw ()
  {
    for (unsigned i = 0; i < outputSize (); ++i)
      {
        gradient_t grad = gradient (argument, i);
        for (unsigned j = 0; j < inputSize (); ++j)
          jacobian (i, j) = grad[j];
      }
  }

  std::ostream&
  DifferentiableFunction::print (std::ostream& o) const throw ()
  {
    if (getName ().empty ())
      return o << "Differentiable function";
    else
      return o << getName () << " (derivable function)";
  }

} // end of namespace roboptim
