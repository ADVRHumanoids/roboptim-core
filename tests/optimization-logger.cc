// Copyright (C) 2014 by Benjamin Chretien, CNRS-LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>

#include <boost/make_shared.hpp>
#include <boost/mpl/vector.hpp>
#include <boost/filesystem.hpp>
#include <boost/xpressive/xpressive.hpp>

#include <roboptim/core/io.hh>
#include <roboptim/core/solver-factory.hh>
#include <roboptim/core/optimization-logger.hh>
#include <roboptim/core/differentiable-function.hh>

using namespace roboptim;

// Define a simple function (non-differentiable)
struct F1 : public Function
{
  typedef Function parent_t;

  F1 () : parent_t (4, 1, "a * d * (a + b + c) + d\nDummy constraint")
  {}

  void impl_compute (result_ref result, const_argument_ref x)
    const
  {
    result (0) = x[0] * x[3] * (x[0] + x[1] + x[2]) + x[3];
  }

  // No gradient, hessian.
};

// Define a simple function (sparse differentiable)
struct F2 : public DifferentiableSparseFunction
{
  typedef DifferentiableSparseFunction parent_t;

  F2 () : parent_t (4, 1, "a + b + c + d\nDummy constraint")
  {}

  void impl_compute (result_ref result, const_argument_ref x)
    const
  {
    result (0) = x[0] + x[1] + x[2] + x[3];
  }

  void impl_gradient (gradient_ref grad, const_argument_ref, size_type)
    const
  {
    grad.insert (0) = 1;
    grad.insert (1) = 1;
    grad.insert (2) = 1;
    grad.insert (3) = 1;
  }
};

bool findRegex (const std::string& text, const std::string& r)
{
  boost::xpressive::sregex rex = boost::xpressive::sregex::compile (r);
  return boost::xpressive::regex_search (text, rex);
}

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

// See: http://stackoverflow.com/a/596348/1043187
template <typename>
struct toString
{
  static char const* value () { return "unknown"; }
};

#define DEF_FUNC_TYPE(X)			\
  template <>					\
  struct toString<X>				\
  {						\
    static char const* value () { return #X; }	\
  }

DEF_FUNC_TYPE(F1);
DEF_FUNC_TYPE(F2);

template <>
struct toString<Function>
{
  static char const* value () { return ""; }
};

template <>
struct toString<DifferentiableSparseFunction>
{
  static char const* value () { return "-d-sparse"; }
};

template <typename T>
bool testXOR (const T& a, const T& b)
{
  return !a != !b;
}

template <typename F>
void testLogger
(boost::shared_ptr<boost::test_tools::output_test_stream> output,
 unsigned int requests)
{
  // Specify the solver that will be used.
  typedef Solver<typename F::traits_t> solver_t;

  // Instantiate the function and the problem.
  boost::shared_ptr<F> f = boost::make_shared<F> ();
  typename solver_t::problem_t pb (f);

  // Add a dummy constraint to test logging
  boost::shared_ptr<F> g = boost::make_shared<F> ();
  pb.addConstraint (g, F::makeInterval (-1., 1.));

  // Give a dummy starting point
  typename F::argument_t x (f->inputSize ());
  x.setZero ();
  pb.startingPoint () = x;

  // Instantiate the factory using the dummy solver.
  SolverFactory<solver_t>
    factory (std::string ("dummy")
             + toString<typename F::parent_t>::value ()
             + "-laststate",
             pb);

  // Retrieve a solver instance from the factory.
  solver_t& solver = factory ();

  // Add optimization logger.
  // Note: scoped to trigger the destructor.
  typedef OptimizationLogger<solver_t> logger_t;
  std::stringstream buffer;
  boost::filesystem::path log_path;

  {
    logger_t logger (solver,
        std::string ("/tmp/roboptim-core-tests/optimization-logger-")
        + toString<F>::value (), true, requests);

    // Get the Boost.Variant minimum.
    typename solver_t::result_t res = solver.minimum ();

    // Get the "real" minimum.
    solver.template getMinimum<SolverError> ();

    // Display problem and solver.
    (*output) << pb << std::endl
      << "---" << std::endl
      << solver << std::endl;

    logger.append ("Append test 1");
    logger << "Append test 2"
      << solver.problem () << "\n";

    log_path = logger.logPath ();

    // Test whether the logging directory exists
    BOOST_CHECK (boost::filesystem::exists (log_path));
    // Test whether journal.log exists
    BOOST_CHECK (boost::filesystem::exists (log_path / "journal.log"));
    // Test whether constraint-violation.csv exists
    BOOST_CHECK (! testXOR (logger.isRequested (logger_t::LOG_CONSTRAINT_VIOLATION),
                            boost::filesystem::exists
                              (log_path / "constraint-violation-evolution.csv")));

    // Test whether appended strings are found
    std::ifstream journal ((log_path / "journal.log").string ().c_str ());
    // Load unfinished journal
    buffer << journal.rdbuf();

    BOOST_CHECK (! testXOR (logger.isRequested (logger_t::LOG_X),
                            findRegex (buffer.str (), "- x:\n")));
    BOOST_CHECK (! testXOR (logger.isRequested (logger_t::LOG_COST),
                            findRegex (buffer.str (), "- f\\(x\\):\n")));
    BOOST_CHECK (! testXOR (logger.isRequested (logger_t::LOG_CONSTRAINT_VIOLATION),
                            findRegex (buffer.str (), "- viol_g\\(x\\):\n")));
    BOOST_CHECK (! testXOR (logger.isRequested (logger_t::LOG_TIME),
                            findRegex (buffer.str (),
                                       "Elapsed time since last call:")));
    BOOST_CHECK (! testXOR (logger.isRequested (logger_t::LOG_SOLVER),
                            findRegex (buffer.str (), "Solver:\n")));

    buffer.clear ();
  }

  // Load finalized journal
  std::ifstream final_journal ((log_path / "journal.log").string ().c_str ());
  buffer << final_journal.rdbuf();

  // Note: triggered during the destructor of the logger
  BOOST_CHECK (! testXOR ((logger_t::LOG_TIME & requests) == logger_t::LOG_TIME,
                          findRegex (buffer.str (), " - total elapsed time:")));

  BOOST_CHECK (findRegex (buffer.str (), "Append test 1"));
  BOOST_CHECK (findRegex (buffer.str (), "Append test 2"));
  BOOST_CHECK (findRegex (buffer.str (), "Problem:\n"));
}

BOOST_AUTO_TEST_CASE (plugin)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("optimization-logger");

  typedef OptimizationLogger<Solver<F1::traits_t> > logger_t;
  testLogger<F1> (output, logger_t::FullLogging ());
  (*output) << "---" << std::endl;
  testLogger<F2> (output, logger_t::FullLogging ());
  (*output) << "---" << std::endl;
  testLogger<F1> (output, 0);

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
