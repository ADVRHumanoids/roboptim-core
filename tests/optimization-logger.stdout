Problem:
  a * d * (a + b + c) + d (not differentiable)
  Arguments bounds: (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf)
  Arguments scales: 1, 1, 1, 1
  Number of constraints: 1
  Constraint 0
      a * d * (a + b + c) + d (not differentiable)
      Bounds: (-1, 1)
      Scales: 1
      Initial value: [1](0)
  Starting point: [4](0,0,0,0)
  Starting value: [1](0)
  Infinity value (for all functions): inf
---
Solver:
  Plugin: dummy-laststate
  Problem:
    a * d * (a + b + c) + d (not differentiable)
    Arguments bounds: (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf)
    Arguments scales: 1, 1, 1, 1
    Number of constraints: 1
    Constraint 0
        a * d * (a + b + c) + d (not differentiable)
        Bounds: (-1, 1)
        Scales: 1
        Initial value: [1](0)
    Starting point: [4](0,0,0,0)
    Starting value: [1](0)
    Infinity value (for all functions): inf
  Result:
    Solver error: The dummy solver always fail.
  Parameters:
    dummy-parameter (dummy parameter): 42
    dummy-parameter2 (yet another dummy parameter): 3
    dummy-parameter3 (just a dummy key): ...and a dummy value!
---
Problem:
  a + b + c + d (differentiable function)
  Arguments bounds: (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf)
  Arguments scales: 1, 1, 1, 1
  Number of constraints: 1
  Constraint 0
      a + b + c + d (differentiable function)
      Bounds: (-1, 1)
      Scales: 1
      Initial value: [1](0)
  Starting point: [4](0,0,0,0)
  Starting value: [1](0)
  Infinity value (for all functions): inf
---
Solver:
  Plugin: dummy-d-sparse-laststate
  Problem:
    a + b + c + d (differentiable function)
    Arguments bounds: (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf)
    Arguments scales: 1, 1, 1, 1
    Number of constraints: 1
    Constraint 0
        a + b + c + d (differentiable function)
        Bounds: (-1, 1)
        Scales: 1
        Initial value: [1](0)
    Starting point: [4](0,0,0,0)
    Starting value: [1](0)
    Infinity value (for all functions): inf
  Result:
    Solver error: The dummy solver always fail.
  Parameters:
    dummy-parameter (dummy parameter): 42
    dummy-parameter2 (yet another dummy parameter): 3
    dummy-parameter3 (just a dummy key): ...and a dummy value!
